### ISSUE TITLE
Do not require access to dependencies that aren't being documented

### ISSUE THREAD

**Issue by shs96c** [bazelbuild/stardoc#93]
It's not always practical to add a dependency on the bzl files that come from third party dependencies. `stardoc` should allow documentation to be generated without introspecting deeply into dependencies.

As an example, I use `MavenInfo` from `rules_jvm_external` in my own rules for building java modules. In order to generate docs using stardoc, I must somehow also include source files from `rules_jvm_external`

**Comment by burkpojken** [comment]
I agree with that it would be very good to not have to depend on visibility of third party bzl files that you do not want to generated documentation for.  
We see the same problem when are using skylib features in our bzl files for which we want to generate documentation.
Example we have our custom config settings in config.bzl 
With content:
load(
    "@bazel_skylib//rules:common_settings.bzl",
    "bool_flag",
    "string_flag",
)

def our_rule(
   ......
)
We want to generate documentation for our_rule()
Since the visibility of in the latest version of @bazel_skylib//rules:common_settings.bzl has been changed to
exports_files(
    glob(["*.bzl"]),
    visibility = ["//:__subpackages__"],
)
We cannot use stardoc anymore with the latest version of skylib. It does not feel right to request changes in skylib because of this.

**Comment by brandjon** [comment]
See #69, this should be addressed when we rewrite the doc extraction to grab real objects from the loading phase instead of mocking it in a separate process.

**Comment by brandjon** [comment]
(P4ing this one as opposed to marking it a dup, but the underlying rewrite is actually a P1 FR that we're aiming to work on this quarter and next.)

**Comment by guw** [comment]
I also started to investigate Stardoc for us and quickly ran into this.

Currently unable to build because `rules_pkg` also has their `.bzl` files defined with _limited_ visibility.

**Comment by tmhdgsn** [comment]
has there been any progress on this at all? Currently unable to build because the rule to be documented uses `java_binary/java_library` from rules_java.

'documentation generation failed: File external/rules_java/java/defs.bzl imported '//java/private:native.bzl'

**Comment by tetromino** [comment]
> has there been any progress on this at all? Currently unable to build because the rule to be documented uses `java_binary/java_library` from rules_java.
> 
> 'documentation generation failed: File external/rules_java/java/defs.bzl imported '//java/private:native.bzl'

@tmhdgsn, please ensure that
* your stardoc target has in `"@rules_java//java:rules"` in its transitive `deps`; and
* you are using rules_java 5.3.2 or newer.

Does that fix the problem? If not, please add detailed repro steps and we will try to help debug.

**Comment by tmhdgsn** [comment]
> > has there been any progress on this at all? Currently unable to build because the rule to be documented uses `java_binary/java_library` from rules_java.
> > 'documentation generation failed: File external/rules_java/java/defs.bzl imported '//java/private:native.bzl'
> 
> @tmhdgsn, please ensure that
> 
> * your stardoc target has in `"@rules_java//java:rules"` in its transitive `deps`; and
> * you are using rules_java 5.3.2 or newer.
> 
> Does that fix the problem? If not, please add detailed repro steps and we will try to help debug.

somehow I had managed to miss that target "@rules_java//java:rules", thank you !! all good now.

**Comment by tetromino** [comment]
@hauserx - one approach would be for Bazel to gain a new non-build command or command-line option - perhaps a new option to `bazel dump` or some new flavor of `bazel query` - which will take a label of a .bzl file and dump out a proto that could be transformed into Markdown by Stardoc's renderer.

Open design questions:
* what this command or option should be called
* should it operate on a single .bzl file or multiple ones
* should it output in Stardoc proto format, or do we also want to retrieve some additional loading-phase data for use by other tools
* for rendering to markdown, do we recommend running the Stardoc renderer from the command line, or via the currently private `stardoc_markdown_renderer` rule (which would have to be cleaned up a bit and made public)?
